### ISSUE TITLE
Add possibility to control Ballooning Device

### ISSUE THREAD

**Issue by ratiborusx** [bpg/terraform-provider-proxmox#634]
**Is your feature request related to a problem? Please describe.**
Maybe it wouldn't be too much of a hassle to add control over Ballooning Device?
The only rationale is that it could be useful even if one does not use dynamic/automatic memory allocation, according to PVE docs:
> Even when using a fixed memory size, the ballooning device gets added to the VM, because it delivers useful information such as how much memory the guest really uses. In general, you should leave ballooning enabled, but if you want to disable it (like for debugging purposes), simply uncheck Ballooning Device or set
balloon: 0
in the configuration.

**Describe the solution you'd like**
Introduce additional parameter to 'memory' configuration/map, something like that:
```
memory {
  dedicated = 4096
  balloon   = SOMEVALUE
}
```
I'm not exactly sure if that should be boolean value because from what i see Ballooning Device is enabled by default in any VM with fixed memory allocation and there's no mentioning of this parameter in VM config at all. By unchecking the box in GUI parameter sets to "0" and is kept inside a config. Maybe it should be a default behavior for provider too. For now the only way i was able to keep BD enabled is to set floating memory to the same amount as dedicated, but it's kinda wacky way:
```
  memory {
    dedicated    = 4096
    floating     = 4096
  }
```

```
root@prox-srv1:~# qm config 104 | grep -E 'ball|mem'
balloon: 4096
memory: 4096
```
So basically 'balloon' inside a VM config could be either '0' (disabled), either any integer number meaning minimum memory in that case. And if it's not present in a config at all it means just 'enabled'?
![Clipboard01](https://github.com/bpg/terraform-provider-proxmox/assets/123507924/5bb0f5e4-f4aa-4ea6-8fa6-179b1e20c8bd)

**Describe alternatives you've considered**
Do not forcibly set 'balloon=0' after cloning VM and keep template's configuration.
Probably related to #360 and #566, same problem with legacy's code default values.

```
root@prox-srv1:~# qm config 104 | grep ball
balloon: 0
```

**Comment by bpg** [comment]
Yeah, the memory settings are confusing. Essentially, `floating` **is** `balloon`, no idea why the original plugin author [decided to rename it](https://github.com/bpg/terraform-provider-proxmox/blob/6753582e4b1999fdf2fd9ea0f499c0cd0f7cd64c/proxmox/nodes/vms/vms_types.go#L313).

Changing the default might be tricky though, as we need to maintain backward compatibility. But I agree with you it would be better to follow the Proxmox's defaults, meaning to auto-set balloon = memory for new VMs.

**Comment by ratiborusx** [comment]
Yeah i also think that following Proxmox's defaults as defaults for provider itself sounds sane. I know that it may break existing manifests or trigger update/re-creation of already provisioned resources but probably its better to do it sooner than later (i mean general approximation of provider's default behaviour to that of Proxmox, not just this specific case).
> Essentially, floating is balloon, no idea why the original plugin author [decided to rename it](https://github.com/bpg/terraform-provider-proxmox/blob/6753582e4b1999fdf2fd9ea0f499c0cd0f7cd64c/proxmox/nodes/vms/vms_types.go#L313).
I see now that he renamed it, but still that "floating" argument does not provide ability to just turn on ballooning device without enabling dynamic memory sharing to some extent. This is the minimal configuration using "floating" argument to enable BD. but that obviously enables memory sharing too:
```
  memory {
    dedicated    = 4096
    floating     = 1
  }
```

For now my go-to way is to set "floating" to the same number as "dedicated" - that way i get BD enabled and hopefully do not trigger dynamic memory allocation (because there should be nothing to share). Hopefully, because i don't really know - maybe it still tries to do something about that in vain and introduces some sort of overhead. They only visible "side effect" is a slightly funny view of memory in Hardware GUI (like "4.00 GiB/4.00Gib").

**Comment by otopetrik** [comment]
It might be good idea to keep the provider defaulting to 'no balloon' for now.
Maybe modify the documentation example to include balloon/floating option, something like:
```
memory {
  dedicated = 512
  floating = 512
}
```
(and a note that the total memory of VM is 512 MiB, not 1 GiB)

Setting `floating = 1` seems like weird idea, my understanding is that the balloon driver would try to free all but 1 MiB of guest memory and return it to the host.

Both the Proxmox's `balloon:` and the renamed `floating` are not great names.

Proxmox's `balloon:` value is *not* the size of the balloon - i.e. the memory allocated by the driver in the guest OS and returned to the host. Instead it is the target size of memory used by the VM *when the balloon device driver/service in the guest is in use*. **Larger** `balloon:` value does mean **smaller** balloon and **more memory** in the VM.

The `floating` name is also bad, but at least it is closer to the actual meaning of the value: `total_vm_memory - returned_to_the_host_by_balloon`.

Not sure what the correct name should be. It cannot be `current`, but maybe something like `target_using_balloon` ?

At the moment, dynamic memory does not work with PCIE device passthrough, see [reddit](https://old.reddit.com/r/VFIO/comments/ll4h19/memory_ballooning_current_state_of_memory/gnp37v1/).

There used to be issues with Windows VMs, but it looks like it [should work now](https://pve.proxmox.com/wiki/Dynamic_Memory_Management#Requirements_for_Windows_VM).

The `virtio-mem`, which should allow dynamic memory *and* PCIE device passthrough, is not yet supported in Proxmox.
Over the last year, there was some development, recent version of the patch is [here](https://lists.proxmox.com/pipermail/pve-devel/2023-June/057690.html).
Related changes are already [merged](https://github.com/proxmox/qemu-server/commit/7f8c808772979f274cdfac1dc7264771a3b7a7ae).

It might be worth delaying renaming fields in provider's `memory` block until changes to `memory:` line in Proxmox VM configuration are merged.
That recent Proxmox patch names existing memory value (`dedicated` in the provider) as `current=` and adds `max=`.

So it would be good idea to make sure `floating` is not renamed to `current`.

**Comment by ratiborusx** [comment]
@otopetrik
> Setting floating = 1 seems like weird idea, my understanding is that the balloon driver would try to free all but 1 MiB of guest memory and return it to the host.

It is really a weird idea, in no way i suggested doing that :)
I said that right now i'm using exactly what you're recommending - setting 'floating' to the same amount as 'dedicated' to enable balloon.
> For now my go-to way is to set "floating" to the same number as "dedicated" - that way i get BD enabled and hopefully do not trigger dynamic memory allocation (because there should be nothing to share).

Probably should've given an example of that, yes.

**Comment by ratiborusx** [comment]
I re-read PVE's docs and tinkered a bit more with that stuff and i agree with @otopetrik, we probably should leave it as it is at least for now. It looks like when you create VM in PVE and set 'Memory' it also sets 'Minimum memory' too and that enables balloon. Basically it is the same what we could achieve in provider by setting 'dedicated' and 'floating' to the same number.
![image](https://github.com/bpg/terraform-provider-proxmox/assets/123507924/74ead1f9-ef83-4f76-8b88-91698fad0305)
![image](https://github.com/bpg/terraform-provider-proxmox/assets/123507924/b84d05bd-e311-4984-affa-65b99fd463d3)
![image](https://github.com/bpg/terraform-provider-proxmox/assets/123507924/3c3763a1-cbe0-4fab-a874-e0c23f5ba226)
For me the weird thing when doing this was that in GUI 'Memory' becomes looking a bit weird with double values but now i suspect it because provider not exactly 1:1 emulates Proxmox's behaviour on that. Probably it is that provider always sets 'balloon' parameter inside VM config even when 'Memory' and 'Minimum memory' is identical (or 'dedicated' and 'floating' in provider's terms):

![image](https://github.com/bpg/terraform-provider-proxmox/assets/123507924/9b128428-709d-43a9-b16c-e90ff648b37b)
```
root@prox-srv1:~# qm config 104 | grep -E 'balloon:|memory:'
balloon: 4096
memory: 4096
```
